prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -1085,34 +992,86 @@ void rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
 		rxr_pkt_entry_release_rx(ep, pkt_entry);
 		return;
 	}
-	
-	if (pkt_entry->addr == FI_ADDR_NOTAVAIL) {
-		if (pkt_type >= RXR_REQ_PKT_BEGIN && rxr_pkt_req_raw_addr(pkt_entry)) {
+
+	/*
+	 * Attempt to inject or infer peer address if not available. This usually
+	 * happens when the endpoint receives the first packet from a new peer.
+	 * There is an edge case for EFA endpoint - the device might lose the address
+	 * handle of a known peer due to a firmware bug and return FI_ADDR_NOTAVAIL.
+	 * The provider needs to infer the address by searching the reverse AV table/map,
+	 * and re-insert the address to the AV if the peer address is found.
+	 */
+	if (pkt_entry->addr == FI_ADDR_NOTAVAIL && pkt_type >= RXR_REQ_PKT_BEGIN) {
```

Comment:
Do not limit the pkt_type here. the `efadv_wc_read_ah` should be available for any pkt type.

Reply:
Do we want to skip it for shm endpoint though?

I did this check to maintain backward compatibility. The current logic only looks for raw address in packet header under this condition.

Reply:
No, for shm. We still want to limit packet type. I think the logic is getting really complicated here, so it is better to have a function for shm and two functions for EFA (under different value of HAVE_EFADV_CQ_EX_SUPPORT).